File: src/main/java/io/github/jakejmattson/touchcontrol/utils/TouchableGroup.java
package io.github.jakejmattson.touchcontrol.utils;

import io.github.jakejmattson.touchcontrol.touchables.Touchable;
import org.opencv.core.*;

import java.util.*;

/**
 * Apply the same function over many Touchable objects.
 *
 * @author JakeJMattson
 */
public class TouchableGroup
{
	/**
	 * A List of all components in the group
	 */
	private final ArrayList<Touchable> components;

	public TouchableGroup(Touchable... components)
	{
		this.components = new ArrayList<>(Arrays.asList(components));
		checkCollision();
	}

	public void setColor(Scalar color)
	{
		for (Touchable component : components)
			component.setColor(color);
	}

	public void addComponent(Touchable component)
	{
		components.add(component);
		checkCollision();
	}

	public void removeComponent(Touchable component)
	{
		components.remove(component);
	}

	public List<Touchable> getComponents()
	{
		return components;
	}

	public void updateDetectionPoint(Mat filteredImage)
	{
		for (Touchable component : components)
			component.updateDetectionPoint(filteredImage);
	}

	public void drawOnto(Mat rawImage)
	{
		for (Touchable component : components)
			component.drawOnto(rawImage);
	}

	public void performAction()
	{
		for (Touchable component : components)
			component.performAction();
	}

	/**
	 * Check whether or not any of the components in
	 * the group overlaps (collides) with another.
	 * Note: This has no return. It is only to print warnings.
	 */
	private void checkCollision()
	{
		for (int i = 0; i < components.size(); i++)
			for (int j = i + 1; j < components.size(); j++)
			{
				//Get components dimensions
				Rect dim1 = components.get(i).getDimensions();
				Rect dim2 = components.get(j).getDimensions();

				//Check if two components overlap
				boolean overlaps = !(dim1.y + dim1.height <= dim2.y
						|| dim1.y >= dim2.y + dim2.height
						|| dim1.x + dim1.width <= dim2.x
						|| dim1.x >= dim2.x + dim2.width);

				//Print warning
				if (overlaps)
					System.out.println("Touchable collision warning in " + super.toString() +
							" between component " + i + " and component " + j);
			}
	}

	/*
	 * (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString()
	{
		String NEWLINE = System.lineSeparator();
		StringBuilder groupData = new StringBuilder();

		groupData.append("Touchable objects in group (").append(super.toString()).append("): ")
				.append(components.size()).append(NEWLINE).append(NEWLINE);

		for (int i = 0; i < components.size(); i++)
			groupData.append("(").append(i).append(")").append(components.get(i).toString()).append(NEWLINE);

		return groupData.toString();
	}
}

File: src/main/java/io/github/jakejmattson/touchcontrol/utils/TouchableGroup.kt
package io.github.jakejmattson.touchcontrol.utils

import io.github.jakejmattson.touchcontrol.touchables.Touchable
import kotlinx.coroutines.*
import org.opencv.core.*
import java.util.*
import kotlin.collections.ArrayList

/**
 * Apply the same function over many Touchable objects.
 *
 * @author JakeJMattson
 */
class TouchableGroup(vararg components: Touchable) {
	/**
	 * A List of all components in the group
	 */
	private val components: ArrayList<Touchable> = ArrayList(Arrays.asList(*components))

	fun addComponent(component: Touchable) = components.add(component)

	fun removeComponent(component: Touchable) = components.remove(component)

	fun getComponents() = components

	fun setColor(color: Scalar) = components.forEach { it.color = color }

	fun updateDetectionPoint(filteredImage: Mat) =
		runBlocking {
			val jobs = ArrayList<Job>()

			components.forEach {
				jobs.add(
					GlobalScope.launch { it.updateDetectionPoint(filteredImage) }
				)
			}

			jobs.joinAll()
		}

	fun drawOnto(rawImage: Mat) =
		runBlocking {
			val jobs = ArrayList<Job>()

			components.forEach {
				jobs.add(
					GlobalScope.launch { it.drawOnto(rawImage) }
				)
			}

			jobs.joinAll()
		}

	fun performAction() =
		runBlocking {
			val jobs = ArrayList<Job>()

			components.forEach {
				jobs.add(
					GlobalScope.launch { it.performAction() }
				)
			}

			jobs.joinAll()
		}

	/**
	 * Check whether or not any of the components in
	 * the group overlaps (collides) with another.
	 *
	 * @return Whether or not the group is collision-free
	 */
	private fun checkCollision(): Boolean {

		var hasNoCollision = true

		for (i in components.indices)
			for (j in i + 1 until components.size) {

				val dim1 = components[i].dimensions
				val dim2 = components[j].dimensions

				//Check if two components overlap
				val overlaps = !(dim1.y + dim1.height <= dim2.y
						|| dim1.y >= dim2.y + dim2.height
						|| dim1.x + dim1.width <= dim2.x
						|| dim1.x >= dim2.x + dim2.width)

				if (overlaps) {
					hasNoCollision = false
					
					println("Touchable collision warning in ${super.toString()} between component $i and component $j")
				}
			}

		return hasNoCollision
	}

	override fun toString(): String {
		var groupData = "Touchable objects in group (${super.toString()}): ${components.size}\n\n"

		components.forEachIndexed { index, component ->
			groupData += "($index)$component\n"
		}

		return groupData
	}
}